tracer/test/test-applications/integrations/Samples.DataStreams.Kafka/Program.cs

```diff
@@ -1,104 +1,234 @@
-}
+    Console.WriteLine($"Creating consumers...");
+    var fanInConsumer = Consumer.Create(topic1, "fan-in-consumer", HandleFanIn);
+    var finalConsumer = Consumer.Create(topic2, "topic-2-consumer", HandleTopic2);
 
-void HandleAndProduceToTopic2(ConsumeResult<string, string> consumeResult)
-    => HandleAndProduce(consumeResult, topic2);
+    Console.WriteLine("Starting consumers...");
+    var cts = new CancellationTokenSource();
+    var consumeTasks = new Task[2];
+    consumeTasks[0] = Task.Run(() => finalConsumer.Consume(cts.Token));
+    consumeTasks[1] = Task.Run(() => fanInConsumer.Consume(cts.Token));
+    LogWithTime("Finished starting consumers");
 
-void HandleAndProduceToTopic3(ConsumeResult<string, string> consumeResult)
-    => HandleAndProduce(consumeResult, topic3);
+    Console.WriteLine($"Producing messages");
+    Producer.Produce(topic1, numMessages: 3, config, handleDelivery: true, isTombstone: false);
+    LogWithTime("Finished producing messages");
 
-void HandleAndProduce(ConsumeResult<string,string> consumeResult, string produceToTopic)
-{
-    Handle(consumeResult);
-    
-    Console.WriteLine($"Producing to {produceToTopic}");
-    Producer.Produce(produceToTopic, numMessages: 1, config, handleDelivery: true, isTombstone: false);
+    // Wait for all messages to be consumed
+    // This assumes that the topics all start empty, and ultimately 2 messages end up consumed from topic 2
+    Console.WriteLine($"Waiting for final consumption...");
+    var deadline = DateTime.UtcNow.AddSeconds(30);
+    while (true)
+    {
+        var consumed = Volatile.Read(ref topic2ConsumeCount);
+        if (consumed >= 2)
+        {
+            Console.WriteLine($"All messages produced and consumed");
+            break;
+        }
+
+        if (DateTime.UtcNow > deadline)
+        {
+            Console.WriteLine($"Exiting consumer: did not consume all messages: {consumed}");
+            break;
+        }
+
+        await Task.Delay(1000);
+    }
+
+    LogWithTime("Finished waiting for messages");
+
+    Console.WriteLine($"Waiting for graceful exit...");
+    cts.Cancel();
+
+    await Task.WhenAny(
+        Task.WhenAll(consumeTasks),
+        Task.Delay(TimeSpan.FromSeconds(5)));
+
+    LogWithTime("Shut down complete");
+
+
+    void HandleTopic2(ConsumeResult<string, string> consumeResult)
+    {
+        using var s = SampleHelpers.CreateScopeWithPropagation(
+            "kafka.consume",
+            consumeResult.Message.Headers,
+            ConsumerBase.ExtractValues);
+        Handle(consumeResult);
+
+        var consumeCount = Interlocked.Increment(ref topic2ConsumeCount);
+        Console.WriteLine($"Consumed message {consumeCount} in Topic({topic2})");
+    }
+
+    void HandleFanIn(ConsumeResult<string, string> consumeResult)
+    {
+        Handle(consumeResult);
+
+        _fanInMessages.Add(consumeResult);
+        if (_fanInMessages.Count == 3)
```

Comment:
Is it possible that two consumers acting on their own threads could both Add at around the same time and then both enter the if block because they both read a count of 3? If that's possible, that could might add flakiness by producing additional data streams points